### PR TITLE
Game creation and joining

### PR DESCRIPTION
**Pull request by FuzzyLogic123**
Rory's changes plus Post api endpoint

### CODE REVIEW

File: client/src/loaders/index.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+export interface NewGameLoaderProps {
+  request: ReactRotuterRequest;
+}
+
+export const newGameLoader = async ({ request }: NewGameLoaderProps) => {
+  const url = new URL(request.url);
+
+  const gameCode = url.searchParams.get("gameCode");
+  const qrCode = url.searchParams.get("qrCode");
```
Comment (FuzzyLogic123) [inline client/src/loaders/index.ts:11]: Don't think we need this and remove comment

File: client/src/pages/GameLobby.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  //     })
+  // }, [socket]);
+
+  return (
+    <div>
+      <div>
+        <DisplayLogo />
+      </div>
+      <h1 className="text-white font-bold mt-6 uppercase">
+        {" "}
```
Comment (FuzzyLogic123) [inline client/src/pages/GameLobby.tsx:32]: Quotes are pointless

File: client/src/pages/HostGame.tsx [hunk 3]
```diff
@@ -1,20 +1,41 @@
...
 
+  const [loading, setLoading] = useState(false);
   const createGame = () => {
     socket.emit("CREATE_GAME", hostName);
+    setLoading(true);
   };
 
+  const [gameCode, setGameCode] = useState("");
+  const [qrCode, setQrCode] = useState("");
+  socket.on("GAME_CREATED", (gameCode, qrCode) => {
```
Comment (FuzzyLogic123) [inline client/src/pages/HostGame.tsx:20]: The socket won't need to return a qr code/ don't need qr code state yet

File: client/src/pages/HostGame.tsx [hunk 4]
```diff
@@ -1,20 +1,41 @@
...
+    // joined the room etc
+    setGameCode(gameCode);
+    setQrCode(qrCode);
+  });
+
+  useEffect(() => {
+    if (loading) {
+      console.log("fafa");
```
Comment (FuzzyLogic123) [inline client/src/pages/HostGame.tsx:28]: remove fafa lol

File: client/src/pages/HostGame.tsx [hunk 5]
```diff
@@ -1,20 +1,41 @@
...
+      navigate(`../${GAME_LOBBY_PATH}?gameCode=${gameCode}&qrCode=${qrCode}`);
```
Comment (FuzzyLogic123) [inline client/src/pages/HostGame.tsx:29]: remove qr param

File: client/src/pages/HostGame.tsx [hunk 6]
```diff
@@ -30,12 +51,18 @@ const HostGame = () => {
       </div>
 
       <div className="w-screen h-12 mt-8">
-        <button
-          className="text-white bg-primary text-2xl font-bold w-1/3 rounded-xl h-full"
-          onClick={createGame}
-        >
-          Create Game
-        </button>
+        {loading ? (
```
Comment (FuzzyLogic123) [inline client/src/pages/HostGame.tsx:54]: Put this ternary on the child inside the same button. If we want to stop the user activating on click event when its loading use the `disabled={loading}` attribute on the button.

File: client/src/pages/JoinGame.tsx [hunk 7]
```diff
@@ -11,6 +13,15 @@ const JoinGame = () => {
     socket.emit("JOIN_GAME", gameCode, playerName);
```
Comment (FuzzyLogic123) [inline client/src/pages/JoinGame.tsx:13]: Should this be in a useEffect too? Also don't we need game code from the URL params?

File: client/src/pages/JoinGame.tsx [hunk 8]
```diff
@@ -11,6 +13,15 @@ const JoinGame = () => {
...
   };
 
+  const [gameCodeCreated, setGameCode] = useState("");
+  const [player, setPlayer] = useState<PlayerAttributes>();
+  useEffect(() => {
+    socket.on("JOINED_GAME", (player, gameCode) => {
```
Comment (FuzzyLogic123) [inline client/src/pages/JoinGame.tsx:19]: We don't need to recieve the gameCode from the socket right? We have access to it in the component state.

File: client/src/pages/JoinGame.tsx [hunk 9]
```diff
@@ -11,6 +13,15 @@ const JoinGame = () => {
     socket.emit("JOIN_GAME", gameCode, playerName);
   };
 
+  const [gameCodeCreated, setGameCode] = useState("");
```
Comment (FuzzyLogic123) [inline client/src/pages/JoinGame.tsx:16]: Whats the difference between the two game code states. Also we need to name the setter function as the same as the state with "set" prefixed.

File: server/server.ts [hunk 10]
```diff
@@ -18,14 +23,75 @@ const io = new Server<Events>(server, {
...
+    // generated gameRoomCode (defaulted atm)
+    const gameCode = "000000";
+    gamesMap.set(gameCode, game);
+
+    // socket.join(hostRoomName(gameCode));
+
+    io.to(host.socketId).emit("GAME_CREATED", gameCode, "000000");
   });
 
+  // app.post('/create-game', async (req, res) => {
```
Comment (FuzzyLogic123) [inline server/server.ts:74]: We can remove this now

File: client/src/pages/JoinGame.tsx (hunk 7 above)
Comment (RoryT-U) [reply client/src/pages/JoinGame.tsx:13]: This is sending the JOIN_GAME even to the server, so we need to input the game code that a Player is trying to join too.

**Review by RoryT-U — APPROVED** [review]

**Pull request merged**